### PR TITLE
CDI-625 Make it clear when exactly are context init/destroy events fired

### PR DESCRIPTION
**Pull request by mkouba**
- add @BeforeDestroyed and @AfterDestroyed
- deprecate @Destroyed

### CODE REVIEW

File: api/src/main/java/javax/enterprise/context/BeforeDestroyed.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * The scope for which to observe destruction
+     */
+    Class<? extends Annotation> value();
+
+    /**
+     * Supports inline instantiation of the {@link BeforeDestroyed} qualifier.
+     */
+    public final static class Literal extends AnnotationLiteral<BeforeDestroyed> implements BeforeDestroyed {
+
+        public static final Literal REQUEST = of(RequestScoped.class);
```
Comment (johnament) [inline api/src/main/java/javax/enterprise/context/BeforeDestroyed.java]: do all of these need to be present? Seem duplicated.

Comment (mkouba) [inline api/src/main/java/javax/enterprise/context/BeforeDestroyed.java]: Nope. This one is for `@BeforeDestroyed(RequestScoped.class)`. The one in `Initialized` is for `@Initialized(RequestScoped.class)` and so on...

**Comment by tremes** [comment]
Looks good.

File: api/src/main/java/javax/enterprise/context/Destroyed.java [hunk 2]
```diff
@@ -33,13 +33,12 @@
 
 /**
  * <p>
- * The <code>@Destroyed</code> qualifier.
+ * This qualifier is deprecated from CDI 2.0 and {@link BeforeDestroyed} or {@link AfterDestroyed} should be used instead.
```
Comment (manovotn) [inline api/src/main/java/javax/enterprise/context/Destroyed.java]: Shouldn't we add a note that the new `@AfterDestroyed` behaves equally to this deprecated one?

Comment (mkouba) [inline api/src/main/java/javax/enterprise/context/Destroyed.java]: Yep, makes sense.

**Comment by manovotn** [comment]
I think it's good now, +1

**Comment by antoinesd** [comment]
Ok for me but I just have a backward compatibility question. Shouldn't we keep mentioning @Destroyed as being deprecated in the spec like we did with @New ?

**Comment by mkouba** [comment]
@antoinesd Do you mean the spec text?

**Comment by antoinesd** [comment]
@mkouba. Yes, we did that for [@New](http://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#new).
I'm not a big fan of letting deprecated features in the spec, but I'm afraid we don't have choice... wdyt ?

**Comment by mkouba** [comment]
If it's not required by some JCP rule I would remove @Destroyed from the spec text.

**Comment by manovotn** [comment]
I think we shouldn't mention it. That would just bloat the spec.
We still remain backward compatible and `@Destroyed` will work for those who have it in code now. Plus the new annotation is nothing far-fetched from the old one - it is easy to figure it out.

**Comment by antoinesd** [comment]
Guys I'm not saying I **want** it to stay in the doc, I'm asking if it's allowed to remove it. I just sent a mail to Linda to be sure. Hope she'll tell we can do it...

**Comment by antoinesd** [comment]
Found: https://java.net/projects/javaee-spec/pages/CompatibilityRequirements
point 6 deals with deprecated methods (not exactly our use case, but very close):

> 6. An expert group may decide to designate a method or interface as being "deprecated" when its use is no longer recommended or when the expert group considers that a different interface or method should be used instead. This can be stated in the specification and in the relevant javadoc. However the @Deprecated annotation must not be used as this is noisy and annoying for users. The only exception to this is if there is a serious security or functional defect in the method or interface.

wdyt ?

**Comment by manovotn** [comment]
Pretty much close to our case. 

> This _**can**_ be stated in the specification

The way I understand it, it's up to us.

**Comment by mkouba** [comment]
@cdi-spec/eg @johnament PR changed - follows EG decision.

**Comment by antoinesd** [comment]
LGTM

**Pull request merged**